hoptimator-operator/src/main/java/com/linkedin/hoptimator/operator/subscription/SubscriptionReconciler.java

```diff
@@ -56,38 +58,80 @@ public Result reconcile(Request request) {
     RequestEnvironment env = new RequestEnvironment(request);
     Resource.TemplateFactory templateFactory = new Resource.SimpleTemplateFactory(env);
 
+    Result result = new Result(true, operator.pendingRetryDuration());
     try {
       V1alpha1Subscription object = operator.<V1alpha1Subscription>fetch(SUBSCRIPTION, namespace,
         name);
-  
+ 
       if (object ==  null) {
         log.info("Object {}/{} deleted, skipping.", namespace, name);
         return new Result(false);
       }
-
-      V1OwnerReference ownerReference = new V1OwnerReference();
-      ownerReference.kind(object.getKind());
-      ownerReference.name(object.getMetadata().getName());
-      ownerReference.apiVersion(object.getApiVersion());
-      ownerReference.uid(object.getMetadata().getUid());
-   
-      Pipeline pipeline = pipeline(object);
-      boolean ready = pipeline.resources().stream()
-        .map(x -> operator.applyResource(x, ownerReference, templateFactory)).allMatch(x -> x);
+      
+      String kind = object.getKind();
 
       V1alpha1SubscriptionStatus status = object.getStatus();
       if (status == null) {
         status = new V1alpha1SubscriptionStatus();
+        object.setStatus(status);
       }
-      status.setReady(ready);
 
-      operator.apiFor(SUBSCRIPTION).updateStatus(object, x -> object.getStatus());
+      // We deploy in three phases:
+      // 1. Plan a pipeline, and write the plan to Status.
+      // 2. Deploy the pipeline per plan.
+      // 3. Verify readiness of the entire pipeline.
+      // Each phase should be a separate reconcilation loop to avoid races.
+      // TODO: We should disown orphaned resources when the pipeline changes.
+      if (status.getSql() == null || !status.getSql().equals(object.getSpec().getSql())) {
+        // Phase 1
+        log.info("Planning a new pipeline for {}/{} with SQL `{}`...", kind, name, object.getSpec().getSql());
+
+        Pipeline pipeline = pipeline(object);
+        status.setResources(pipeline.resources().stream()
+          .map(x -> x.render(templateFactory))
+          .collect(Collectors.toList()));
+
+        status.setSql(object.getSpec().getSql());
+        status.setReady(null);  // null indicates that pipeline needs to be deployed
+        status.setMessage("Planned.");
+      } else if (status.getReady() == null && status.getResources() != null) {
+        // Phase 2
+        log.info("Deploying pipeline for {}/{}...", kind, name);
+
+        boolean deployed = status.getResources().stream()
+          .map(x -> apply(x, object)).allMatch(x -> x);
+
+        if (deployed) {
+          status.setReady(false);
+          status.setMessage("Deployed.");
+        } else {
+          return new Result(true, operator.failureRetryDuration());
+        }
+      } else {
+        log.info("Checking status of pipeline for {}/{}...", kind, name);
+
+        boolean ready = status.getResources().stream()
+          .map(x -> checkStatus(x)).allMatch(x -> x);
```

Comment:
nit (optional): `.map(...)` is unnecessary and is equivalent to "just" `allMatch(...)`:

```java
.allMatch(this::checkStatus);
```